src/drivers/dw/dma.c

```diff
@@ -639,13 +639,19 @@ static int dw_dma_stop(struct dma *dma, int channel)
 
 	spin_lock_irq(&dma->lock, flags);
 
-	ret = poll_for_register_delay(dma_base(dma) + DW_DMA_CHAN_EN,
-				      CHAN_MASK(channel), 0,
+	dw_write(dma, DW_CFG_LOW(channel), chan->cfg_lo | DW_CFG_CH_SUSPEND |
+		 DW_CFG_CH_DRAIN);
+
+	ret = poll_for_register_delay(dma_base(dma) + DW_CFG_LOW(channel),
+				      DW_CFG_CH_FIFO_EMPTY,
+				      DW_CFG_CH_FIFO_EMPTY,
 				      PLATFORM_DMA_TIMEOUT);
 	if (ret < 0)
 		trace_dwdma_error("dw-dma: %d channel %d timeout",
 				  dma->plat_data.id, channel);
 
+	dw_write(dma, DW_DMA_CHAN_EN, CHAN_DISABLE(channel));
```

Comment:
As you mentioned above the dw-dma will auto stop after we wait. Why need to disable it again?

Reply:
Everything is explained in commit message. Wihtout explicit stop you can potentially wait up to 1 ms, which for sure will cause IPC timeout.